fastval.go

```diff
@@ -16,34 +16,85 @@ import (
 type ValueType int
 
 // This must be in comparison precedence order!
+// Mirrors as closely as possible with N1QL Collate order
 const (
 	InvalidValue ValueType = iota
 	MissingValue
-	IntValue
+	NullValue
+	FalseValue
+	TrueValue
+	// Numerics:
 	UintValue
-	JsonIntValue
 	JsonUintValue
+	IntValue
+	JsonIntValue
 	FloatValue
 	JsonFloatValue
+	// String types
 	StringValue
 	BinStringValue
 	JsonStringValue
-	RegexValue
-	PcreValue
-	BinaryValue
-	NullValue
-	TrueValue
-	FalseValue
+	// Time can only be implicitly converted from a specific string form
+	TimeValue
+	// Array Obj
 	ArrayValue
 	ObjectValue
-	TimeValue
+	// Binary types
+	BinaryValue
+	RegexValue
+	PcreValue
 )
 
+// Implicit Conversion Table
+// Keyed by type, and then a list of potentially convertible target datatype with
+// "true" for definitely convertible and "false" for potentially convertible
+// If a ValueType does not exist in this table, that means there is no possible target conversion
+var ImplicitConvTable = map[ValueType]map[ValueType]bool{
```

Comment:
I am trying to understand the need for implicit conversion. Can it be simplified by going with strict types where filter type should match with document?

Reply:
That would simplify from a software development perspective, but could turn into many customer issues where the filter doesn't work according to their expectation.